"Couldn't find the binary sh" with yarn 1.5.1

<!-- *Before creating an issue please make sure you are using the latest version of yarn.* -->

**Do you want to request a *feature* or report a *bug*?**
<!-- Is the feature a substantial feature request? Please use https://github.com/yarnpkg/rfcs -->

**What is the current behavior?**
I get this installing our yarn packages., With yarn 1.3.2 it was working fine. This is on a linux box. On my windows box it works fine.

```
2018-03-07 11:43:17 gobuilds.Compile : [INFO] yarn install v1.5.1
2018-03-07 11:43:17 gobuilds.Compile : [INFO] [1/4] Resolving packages...
2018-03-07 11:43:18 gobuilds.Compile : [INFO] [2/4] Fetching packages...
2018-03-07 11:43:31 gobuilds.Compile : [INFO] info fsevents@1.1.3: The platform "linux" is incompatible with this module.
2018-03-07 11:43:31 gobuilds.Compile : [INFO] info "fsevents@1.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
2018-03-07 11:43:31 gobuilds.Compile : [INFO] [3/4] Linking dependencies...
2018-03-07 11:43:37 gobuilds.Compile : [INFO] [4/4] Building fresh packages...
2018-03-07 11:43:38 gobuilds.Compile : [ERROR] error An unexpected error occurred: "/build/mts/release/sb-14177014/vcac/core/branding/navigation/node_modules/webpack/node_modules/uglifyjs-webpack-plugin: Couldn't find the binary sh".
2018-03-07 11:43:38 gobuilds.Compile : [INFO] info If you think this is a bug, please open a bug report with the information provided in "/build/mts/release/sb-14177014/vcac/core/branding/navigation/yarn-error.log".
2018-03-07 11:43:38 gobuilds.Compile : [INFO] info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
2018-03-07 11:43:38 gobuilds.Compile : [INFO] info This module is OPTIONAL, you can safely ignore this error
2018-03-07 11:43:38 gobuilds.Compile : [ERROR] warning Error running install script for optional dependency: "/build/mts/release/sb-14177014/vcac/core/branding/navigation/node_modules/node-sass: Couldn't find the binary sh"
2018-03-07 11:43:38 gobuilds.Compile : [INFO] info This module is OPTIONAL, you can safely ignore this error
2018-03-07 11:43:38 gobuilds.Compile : [ERROR] warning Error running install script for optional dependency: "/build/mts/release/sb-14177014/vcac/core/branding/navigation/node_modules/uws: Couldn't find the binary sh"
```

**If the current behavior is a bug, please provide the steps to reproduce.**
<!-- If you can, provide a link to a public repository which contains the files necessary to reproduce this. -->

running 'yarn install'. With yarn 1.3.2 it is working fine. All I did was replace to yarn 1.5.1

**What is the expected behavior?**

Expect to install packages

**Please mention your node.js, yarn and operating system version.**

Node 8.9.4
Yarn 1.5.1
Linux

--- a/__tests__/integration.js
+++ b/__tests__/integration.js
@@ -354,10 +354,26 @@ test('yarn run <script> <strings that need escaping>', async () => {
 
   const options = {cwd, env: {YARN_SILENT: 1}};
 
-  const trickyStrings = ['$PWD', '%CD%', '^', '!', '\\', '>', '<', '|', '&', "'", '"', '`', '  '];
+  const trickyStrings = ['$PWD', '%CD%', '^', '!', '\\', '>', '<', '|', '&', "'", '"', '`', '  ', '()'];
   const [stdout] = await runYarn(['stringify', ...trickyStrings], options);
 
   expect(stdout.toString().trim()).toEqual(JSON.stringify(trickyStrings));
+});
+
+test('yarn run in path need escaping', async () => {
+  const cwd = await makeTemp('special (chars)');
+
+  await fs.writeFile(path.join(cwd, 'package.json'), '{}');
+  const binDir = path.join(cwd, 'node_modules', '.bin');
+  await fs.mkdirp(binDir);
+  await fs.writeFile(path.join(binDir, 'yolo'), 'echo yolo');
+  await fs.writeFile(path.join(binDir, 'yolo.cmd'), '@ECHO off\necho yolo');
+
+  const options = {cwd, env: {YARN_SILENT: 1}};
+
+  const [stdout] = await runYarn(['yolo'], options);
+
+  expect(stdout.toString().trim()).toEqual('yolo');
 });
 
 test('cache folder fallback', async () => {

--- a/__tests__/integration.js
+++ b/__tests__/integration.js
@@ -366,8 +366,11 @@ test('yarn run in path need escaping', async () => {
   await fs.writeFile(path.join(cwd, 'package.json'), '{}');
   const binDir = path.join(cwd, 'node_modules', '.bin');
   await fs.mkdirp(binDir);
-  await fs.writeFile(path.join(binDir, 'yolo'), 'echo yolo');
-  await fs.writeFile(path.join(binDir, 'yolo.cmd'), '@ECHO off\necho yolo');
+  const executablePath = path.join(binDir, 'yolo');
+  await fs.writeFile(executablePath, 'echo yolo');
+  await fs.chmod(executablePath, 0o755);
+  // For Windows
+  await fs.writeFile(`${executablePath}.cmd`, '@ECHO off\necho yolo');
 
   const options = {cwd, env: {YARN_SILENT: 1}};
 

--- a/__tests__/integration.js
+++ b/__tests__/integration.js
@@ -392,7 +392,7 @@ test('yarn run <script> <strings that need escaping>', async () => {
 
   const options = {cwd, env: {YARN_SILENT: 1}};
 
-  const trickyStrings = ['$PWD', '%CD%', '^', '!', '\\', '>', '<', '|', '&', "'", '"', '`', '  ', '()'];
+  const trickyStrings = ['$PWD', '%CD%', '^', '!', '\\', '>', '<', '|', '&', "'", '"', '`', '  ', '(', ')'];
   const [stdout] = await runYarn(['stringify', ...trickyStrings], options);
 
   expect(stdout.toString().trim()).toEqual(JSON.stringify(trickyStrings));

--- a/src/util/execute-lifecycle-script.js
+++ b/src/util/execute-lifecycle-script.js
@@ -184,18 +184,7 @@ export async function executeLifecycleScript(
 
   await checkForGypIfNeeded(config, cmd, env[constants.ENV_PATH_KEY].split(path.delimiter));
 
-  // get shell
-  let sh = 'sh';
-  let shFlag = '-c';
-
-  let windowsVerbatimArguments = undefined;
-
-  if (customShell) {
-    sh = customShell;
-  } else if (process.platform === 'win32') {
-    sh = process.env.comspec || 'cmd';
-    shFlag = '/d /s /c';
-    windowsVerbatimArguments = true;
+  if (process.platform === 'win32' && (!customShell || customShell === 'cmd')) {
     // handle windows run scripts starting with a relative path
     cmd = fixCmdWinSlashes(cmd);
   }
@@ -219,7 +208,9 @@ export async function executeLifecycleScript(
       }
     };
   }
-  const stdout = await child.spawn(sh, [shFlag, cmd], {cwd, env, stdio, windowsVerbatimArguments}, updateProgress);
+  const stdout = customShell
+    ? await child.spawn(customShell, [cmd], {cwd, env, stdio, windowsVerbatimArguments: true}, updateProgress)
+    : await child.spawn(cmd, [], {cwd, env, stdio, shell: true}, updateProgress);
 
   return {cwd, command: cmd, stdout};
 }

